File: mcp-code-intelligence-python/src/mcp_code_intel/parsers/languages/python_parser.py
"""
KSA-178: Python Language Parser.
Port of mcp-code-intelligence-nodejs/src/parsers/languages/python-parser.ts.
Extracts symbols and relationships from Python AST using tree-sitter.
"""

from __future__ import annotations

import re

from tree_sitter import Parser as TSParser

from ..types import (
    ExtractedRelationship, ExtractedSymbol, ParseResult, SymbolKind,
)
from ..ast_utils import (
    find_nodes, find_first, get_node_text, get_node_range, get_named_child,
    walk_tree, SyntaxNode,
)
from .base_parser import BaseLanguageParser

_PYTHON_BUILTINS = frozenset([
    "print", "len", "range", "str", "int", "float", "list", "dict",
    "set", "tuple", "type", "isinstance", "issubclass", "super",
    "hasattr", "getattr", "setattr", "repr", "bool", "enumerate",
    "zip", "map", "filter", "sorted", "reversed", "any", "all",
    "min", "max", "abs", "round", "open", "id", "hex", "oct",
    "bin", "ord", "chr", "format", "vars", "dir", "help",
    "input", "iter", "next", "slice", "object", "property",
    "staticmethod", "classmethod",
])


class PythonParser(BaseLanguageParser):
    """Extracts symbols and relationships from Python AST."""

    def get_supported_extensions(self) -> list[str]:
        return [".py", ".pyi"]

    def parse(self, source: str, file_path: str) -> ParseResult:
        root = self._parse_tree(source)
        symbols: list[ExtractedSymbol] = []
        relationships: list[ExtractedRelationship] = []
        errors = self._collect_errors(root)

        self._extract_imports(root, source, file_path, relationships)
        self._extract_declarations(root, source, file_path, None, symbols, relationships)

        return ParseResult(symbols=symbols, relationships=relationships, errors=errors)

    # ─── Import Extraction ──────────────────────────────────────────────

    def _extract_imports(
        self, root: SyntaxNode, source: str, file_path: str,
        relationships: list[ExtractedRelationship],
    ) -> None:
        # import_statement: import os, import os.path
        for stmt in find_nodes(root, "import_statement"):
            for name in find_nodes(stmt, "dotted_name"):
                target = get_node_text(name, source)
                relationships.append(ExtractedRelationship(
                    source_symbol="__file__",
                    target_symbol=target,
                    kind="imports",
                    file_path=file_path,
                    line=stmt.start_point[0] + 1,
                ))

        # import_from_statement: from x import y
        for stmt in find_nodes(root, "import_from_statement"):
            module_name = self._extract_module_name(stmt, source)
            is_relative = module_name.startswith(".")

            # Wildcard import
            if find_nodes(stmt, "wildcard_import"):
                relationships.append(ExtractedRelationship(
                    source_symbol="__file__",
                    target_symbol=f"{module_name}.*",
                    kind="imports",
                    file_path=file_path,
                    line=stmt.start_point[0] + 1,
                    metadata={"wildcard": True, **({"relative": True} if is_relative else {})},
                ))
                continue

            # Named imports
            imported_names = find_nodes(stmt, "aliased_import")
            if imported_names:
                for imported in imported_names:
                    children = imported.children
                    if not children:
                        continue
                    name = get_node_text(children[0], source)
                    alias = get_node_text(children[2], source) if len(children) > 2 else None

                    relationships.append(ExtractedRelationship(
                        source_symbol="__file__",
                        target_symbol=f"{module_name}.{name}" if module_name else name,
                        kind="imports",
                        file_path=file_path,
                        line=stmt.start_point[0] + 1,
                        metadata={
                            "from": module_name, "name": name,
                            **({"alias": alias} if alias else {}),
                            **({"relative": True} if is_relative else {}),
                        },
                    ))
            else:
                # Simple identifiers after 'import' keyword
                for name in self._get_imported_identifiers(stmt, source):
                    relationships.append(ExtractedRelationship(
                        source_symbol="__file__",
                        target_symbol=f"{module_name}.{name}" if module_name else name,
                        kind="imports",
                        file_path=file_path,
                        line=stmt.start_point[0] + 1,
                        metadata={"from": module_name, "name": name},
                    ))

    def _extract_module_name(self, stmt: SyntaxNode, source: str) -> str:
        for child in stmt.children:
            if child.type in ("dotted_name", "relative_import"):
                return get_node_text(child, source)
        return ""

    def _get_imported_identifiers(self, stmt: SyntaxNode, source: str) -> list[str]:
        names: list[str] = []
        after_import = False
        for child in stmt.children:
            if child.type == "import":
                after_import = True
                continue
            if after_import and child.type in ("dotted_name", "identifier"):
                names.append(get_node_text(child, source))
        return names

    # ─── Declaration Extraction ─────────────────────────────────────────

    def _extract_declarations(
        self, node: SyntaxNode, source: str, file_path: str,
        parent_name: str | None, symbols: list[ExtractedSymbol],
        relationships: list[ExtractedRelationship],
    ) -> None:
        for child in node.named_children:
            if child.type == "function_definition":
                self._extract_function(child, source, file_path, parent_name, symbols, relationships)
            elif child.type == "class_definition":
                self._extract_class(child, source, file_path, parent_name, symbols, relationships)
            elif child.type == "decorated_definition":
                self._extract_decorated(child, source, file_path, parent_name, symbols, relationships)
            elif child.type == "expression_statement" and not parent_name:
                self._extract_module_variable(child, source, file_path, symbols)

    def _extract_decorated(
        self, node: SyntaxNode, source: str, file_path: str,
        parent_name: str | None, symbols: list[ExtractedSymbol],
        relationships: list[ExtractedRelationship],
    ) -> None:
        for child in node.named_children:
            if child.type == "function_definition":
                self._extract_function(child, source, file_path, parent_name, symbols, relationships, node)
            elif child.type == "class_definition":
                self._extract_class(child, source, file_path, parent_name, symbols, relationships, node)

    # ─── Function Extraction ────────────────────────────────────────────

    def _extract_function(
        self, node: SyntaxNode, source: str, file_path: str,
        parent_name: str | None, symbols: list[ExtractedSymbol],
        relationships: list[ExtractedRelationship],
        decorated_node: SyntaxNode | None = None,
    ) -> None:
        name_node = get_named_child(node, "identifier")
        if not name_node:
            return

        name = get_node_text(name_node, source)
        range_node = decorated_node or node
        start_line, end_line = get_node_range(range_node)

        # Detect async
        preceding_text = source[max(0, node.start_byte - 6):node.start_byte]
        is_async = "async" in preceding_text

        # Extract decorators
        decorators = self._get_decorators(decorated_node or node, source)

        # Determine kind
        kind: SymbolKind = "method" if parent_name else "function"
        if "property" in decorators:
            kind = "property"
        if name == "__init__":
            kind = "constructor"

        # Extract parameters
        params_node = get_named_child(node, "parameters")
        params = get_node_text(params_node, source) if params_node else "()"

        # Extract return type
        return_type = self._extract_return_type(node, source)

        # Modifiers
        modifiers: list[str] = []
        if is_async:
            modifiers.append("async")
        if "staticmethod" in decorators:
            modifiers.append("static")
        if "classmethod" in decorators:
            modifiers.append("classmethod")
        if "abstractmethod" in decorators:
            modifiers.append("abstract")

        is_exported = not name.startswith("_")
        body = get_named_child(node, "block")
        complexity = self._calculate_python_complexity(body) if body else 1
        doc_comment = self._extract_docstring(body, source) if body else None

        symbols.append(ExtractedSymbol(
            name=name,
            kind=kind,
            file_path=file_path,
            start_line=start_line,
            end_line=end_line,
            signature=self._build_func_sig(is_async, name, params, return_type),
            parameters=params,
            return_type=return_type,
            modifiers=modifiers if modifiers else None,
            decorators=decorators if decorators else None,
            parent_name=parent_name,
            is_async=is_async,
            is_exported=is_exported,
            doc_comment=doc_comment,
            complexity=complexity,
        ))

        # Extract calls from body
        if body:
            caller = f"{parent_name}.{name}" if parent_name else name
            self._extract_calls(body, source, file_path, caller, relationships)

        # Extract nested definitions
        if body:
            self._extract_declarations(body, source, file_path, name, symbols, relationships)

        # Decorator relationships
        for dec in (decorators or []):
            relationships.append(ExtractedRelationship(
                source_symbol=f"{parent_name}.{name}" if parent_name else name,
                target_symbol=dec,
                kind="decorates",
                file_path=file_path,
                line=start_line,
            ))

    # ─── Class Extraction ───────────────────────────────────────────────

    def _extract_class(
        self, node: SyntaxNode, source: str, file_path: str,
        parent_name: str | None, symbols: list[ExtractedSymbol],
        relationships: list[ExtractedRelationship],
        decorated_node: SyntaxNode | None = None,
    ) -> None:
        name_node = get_named_child(node, "identifier")
        if not name_node:
            return

        name = get_node_text(name_node, source)
        range_node = decorated_node or node
        start_line, end_line = get_node_range(range_node)
        decorators = self._get_decorators(decorated_node or node, source)

        # Extract base classes
        arg_list = get_named_child(node, "argument_list")
        bases: list[str] = []
        is_protocol = False
        is_abc = False

        if arg_list:
            for arg in arg_list.named_children:
                if arg.type in ("identifier", "attribute"):
                    base_name = get_node_text(arg, source)
                    bases.append(base_name)
                    if base_name == "Protocol":
                        is_protocol = True
                    if base_name in ("ABC", "ABCMeta"):
                        is_abc = True

                    rel_kind = "implements" if is_protocol else "inherits"
                    relationships.append(ExtractedRelationship(
                        source_symbol=name,
                        target_symbol=base_name,
                        kind=rel_kind,
                        file_path=file_path,
                        line=arg.start_point[0] + 1,
                    ))

        kind: SymbolKind = "interface" if is_protocol else "class"
        modifiers: list[str] = []
        if is_abc:
            modifiers.append("abstract")
        if "dataclass" in (decorators or []):
            modifiers.append("dataclass")

        is_exported = not name.startswith("_")
        body = get_named_child(node, "block")
        doc_comment = self._extract_docstring(body, source) if body else None

        symbols.append(ExtractedSymbol(
            name=name,
            kind=kind,
            file_path=file_path,
            start_line=start_line,
            end_line=end_line,
            signature=f"class {name}({', '.join(bases)})" if bases else f"class {name}",
            modifiers=modifiers if modifiers else None,
            decorators=decorators if decorators else None,
            parent_name=parent_name,
            is_exported=is_exported,
            doc_comment=doc_comment,
        ))

        # Extract class body
        if body:
            self._extract_declarations(body, source, file_path, name, symbols, relationships)

        # Decorator relationships
        for dec in (decorators or []):
            relationships.append(ExtractedRelationship(
                source_symbol=name,
                target_symbol=dec,
                kind="decorates",
                file_path=file_path,
                line=start_line,
            ))

    # ─── Module Variable Extraction ─────────────────────────────────────

    def _extract_module_variable(
        self, node: SyntaxNode, source: str, file_path: str,
        symbols: list[ExtractedSymbol],
    ) -> None:
        assignment = get_named_child(node, "assignment")
        if not assignment:
            return

        children = assignment.children
        if not children or children[0].type != "identifier":
            return

        name = get_node_text(children[0], source)
        start_line, end_line = get_node_range(node)
        is_constant = bool(re.match(r"^[A-Z_][A-Z0-9_]*$", name))

        symbols.append(ExtractedSymbol(
            name=name,
            kind="constant" if is_constant else "variable",
            file_path=file_path,
            start_line=start_line,
            end_line=end_line,
            signature=get_node_text(node, source).split("\n")[0].strip()[:200],
            is_exported=not name.startswith("_"),
        ))

    # ─── Call Extraction ────────────────────────────────────────────────

    def _extract_calls(
        self, body: SyntaxNode, source: str, file_path: str,
        caller_name: str, relationships: list[ExtractedRelationship],
    ) -> None:
        call_nodes = find_nodes(body, "call")
        seen: set[str] = set()

        for call in call_nodes:
            children = call.children
            if not children:
                continue
            func_node = children[0]
            func_name = get_node_text(func_node, source)

            if func_name in _PYTHON_BUILTINS:
                continue

            key = f"{caller_name}->{func_name}"
            if key in seen:
                continue
            seen.add(key)

            relationships.append(ExtractedRelationship(
                source_symbol=caller_name,
                target_symbol=func_name,
                kind="calls",
                file_path=file_path,
                line=call.start_point[0] + 1,
            ))

    # ─── Helpers ────────────────────────────────────────────────────────

    def _get_decorators(self, node: SyntaxNode, source: str) -> list[str]:
        decorators: list[str] = []
        for child in node.named_children:
            if child.type == "decorator":
                text = get_node_text(child, source).lstrip("@").split("(")[0].strip()
                decorators.append(text)
        return decorators

    def _extract_return_type(self, node: SyntaxNode, source: str) -> str | None:
        type_node = get_named_child(node, "type")
        if type_node:
            return get_node_text(type_node, source)
        return None

    def _extract_docstring(self, body: SyntaxNode, source: str) -> str | None:
        if not body.named_children:
            return None
        first_child = body.named_children[0]
        if first_child.type != "expression_statement":
            return None
        if not first_child.named_children:
            return None
        expr = first_child.named_children[0]
        if expr.type != "string":
            return None

        text = get_node_text(expr, source)
        text = re.sub(r'^("""|\'\'\')\s*', "", text)
        text = re.sub(r'\s*("""|\'\'\')\s*$', "", text)
        return text.strip()[:500] or None

    def _calculate_python_complexity(self, node: SyntaxNode) -> int:
        complexity = 1
        branch_types = {
            "if_statement", "elif_clause", "for_statement", "while_statement",
            "except_clause", "with_statement", "case_clause", "assert_statement",
        }

        def _enter(n: SyntaxNode) -> None:
            nonlocal complexity
            if n.type in branch_types:
                complexity += 1
            if n.type == "boolean_operator":
                complexity += 1
            if n.type == "conditional_expression":
                complexity += 1
            if n.type in ("list_comprehension", "set_comprehension",
                          "dictionary_comprehension", "generator_expression"):
                complexity += 1

        walk_tree(node, enter=_enter)
        return complexity

    def _build_func_sig(
        self, is_async: bool, name: str, params: str, return_type: str | None,
    ) -> str:
        prefix = "async " if is_async else ""
        ret = f" -> {return_type}" if return_type else ""
        return f"{prefix}def {name}{params}{ret}"[:500]


# Module-level alias for dynamic loading
Parser = PythonParser

File: mcp-code-intelligence-nodejs/src/parsers/languages/python-parser.ts
/**
 * KSA-148: Python Language Parser.
 * Extracts symbols and relationships from Python AST using tree-sitter.
 * Supports: classes (Protocol, ABC, dataclass), functions (async, decorators),
 * imports (all patterns), type hints, complexity calculation.
 */

import type { Parser as ParserType } from 'web-tree-sitter';
import type {
  ILanguageParser, ParseResult, ExtractedSymbol, ExtractedRelationship,
  ParseError, SyntaxNode, SymbolKind, RelationshipKind
} from '../types.js';
import {
  getNodeText, getNodeRange, findNodes, getNamedChild, walkTree, extractDocComment
} from '../ast-utils.js';

export default class PythonParser implements ILanguageParser {
  readonly languageId: string;
  private parser: any;

  constructor(parser: any, languageId: string) {
    this.parser = parser;
    this.languageId = languageId;
  }

  getSupportedExtensions(): string[] {
    return ['.py', '.pyi'];
  }

  parse(source: string, filePath: string): ParseResult {
    const tree = this.parser.parse(source);
    const symbols: ExtractedSymbol[] = [];
    const relationships: ExtractedRelationship[] = [];
    const errors: ParseError[] = [];

    if (tree.rootNode.hasError()) {
      const errorNodes = findNodes(tree.rootNode, 'ERROR');
      for (const node of errorNodes.slice(0, 10)) {
        errors.push({
          message: 'Parse error',
          line: node.startPosition.row + 1,
          column: node.startPosition.column,
        });
      }
    }

    this.extractImports(tree.rootNode, source, filePath, relationships);
    this.extractDeclarations(tree.rootNode, source, filePath, null, symbols, relationships);

    return { symbols, relationships, errors };
  }

  // ─── Import Extraction ───────────────────────────────────────────────

  private extractImports(
    root: SyntaxNode, source: string, filePath: string,
    relationships: ExtractedRelationship[],
  ): void {
    // import_statement: import os, import os.path
    const importStmts = findNodes(root, 'import_statement');
    for (const stmt of importStmts) {
      const names = findNodes(stmt, 'dotted_name');
      for (const name of names) {
        const target = getNodeText(name, source);
        relationships.push({
          sourceSymbol: '__file__',
          targetSymbol: target,
          kind: 'imports',
          filePath,
          line: stmt.startPosition.row + 1,
        });
      }
    }

    // import_from_statement: from x import y
    const fromStmts = findNodes(root, 'import_from_statement');
    for (const stmt of fromStmts) {
      const moduleName = this.extractModuleName(stmt, source);
      const isRelative = moduleName.startsWith('.');

      // Wildcard import
      if (findNodes(stmt, 'wildcard_import').length > 0) {
        relationships.push({
          sourceSymbol: '__file__',
          targetSymbol: `${moduleName}.*`,
          kind: 'imports',
          filePath,
          line: stmt.startPosition.row + 1,
          metadata: { wildcard: true, ...(isRelative && { relative: true }) },
        });
        continue;
      }

      // Named imports (aliased_import nodes)
      const importedNames = findNodes(stmt, 'aliased_import');
      if (importedNames.length > 0) {
        for (const imported of importedNames) {
          const nameNode = imported.child(0);
          if (!nameNode) continue;
          const name = getNodeText(nameNode, source);
          const aliasNode = imported.childCount > 2 ? imported.child(2) : null;
          const alias = aliasNode ? getNodeText(aliasNode, source) : undefined;

          relationships.push({
            sourceSymbol: '__file__',
            targetSymbol: moduleName ? `${moduleName}.${name}` : name,
            kind: 'imports',
            filePath,
            line: stmt.startPosition.row + 1,
            metadata: {
              from: moduleName,
              name,
              ...(alias && { alias }),
              ...(isRelative && { relative: true }),
            },
          });
        }
      } else {
        // Simple identifiers after 'import' keyword
        const identifiers = this.getImportedIdentifiers(stmt, source);
        for (const name of identifiers) {
          relationships.push({
            sourceSymbol: '__file__',
            targetSymbol: moduleName ? `${moduleName}.${name}` : name,
            kind: 'imports',
            filePath,
            line: stmt.startPosition.row + 1,
            metadata: { from: moduleName, name },
          });
        }
      }
    }
  }

  private extractModuleName(stmt: SyntaxNode, source: string): string {
    for (let i = 0; i < stmt.childCount; i++) {
      const child = stmt.child(i);
      if (!child) continue;
      if (child.type === 'dotted_name' || child.type === 'relative_import') {
        return getNodeText(child, source);
      }
    }
    return '';
  }

  private getImportedIdentifiers(stmt: SyntaxNode, source: string): string[] {
    const names: string[] = [];
    let afterImport = false;
    for (let i = 0; i < stmt.childCount; i++) {
      const child = stmt.child(i);
      if (!child) continue;
      if (child.type === 'import') { afterImport = true; continue; }
      if (afterImport && (child.type === 'dotted_name' || child.type === 'identifier')) {
        names.push(getNodeText(child, source));
      }
    }
    return names;
  }

  // ─── Declaration Extraction ──────────────────────────────────────────

  private extractDeclarations(
    node: SyntaxNode, source: string, filePath: string,
    parentName: string | null, symbols: ExtractedSymbol[],
    relationships: ExtractedRelationship[],
  ): void {
    for (let i = 0; i < node.namedChildCount; i++) {
      const child = node.namedChild(i);
      if (!child) continue;

      switch (child.type) {
        case 'function_definition':
          this.extractFunction(child, source, filePath, parentName, symbols, relationships);
          break;
        case 'class_definition':
          this.extractClass(child, source, filePath, parentName, symbols, relationships);
          break;
        case 'decorated_definition':
          this.extractDecorated(child, source, filePath, parentName, symbols, relationships);
          break;
        case 'expression_statement':
          if (!parentName) {
            this.extractModuleVariable(child, source, filePath, symbols);
          }
          break;
      }
    }
  }

  private extractDecorated(
    node: SyntaxNode, source: string, filePath: string,
    parentName: string | null, symbols: ExtractedSymbol[],
    relationships: ExtractedRelationship[],
  ): void {
    for (let i = 0; i < node.namedChildCount; i++) {
      const child = node.namedChild(i);
      if (!child) continue;
      if (child.type === 'function_definition') {
        this.extractFunction(child, source, filePath, parentName, symbols, relationships, node);
      } else if (child.type === 'class_definition') {
        this.extractClass(child, source, filePath, parentName, symbols, relationships, node);
      }
    }
  }

  // ─── Function Extraction ─────────────────────────────────────────────

  private extractFunction(
    node: SyntaxNode, source: string, filePath: string,
    parentName: string | null, symbols: ExtractedSymbol[],
    relationships: ExtractedRelationship[],
    decoratedNode?: SyntaxNode,
  ): void {
    const nameNode = getNamedChild(node, 'identifier');
    if (!nameNode) return;

    const name = getNodeText(nameNode, source);
    const range = getNodeRange(decoratedNode || node);

    // Detect async — check preceding text for 'async' keyword
    const precedingText = source.substring(Math.max(0, node.startIndex - 6), node.startIndex);
    const isAsync = precedingText.includes('async');

    // Extract decorators
    const decorators = this.getDecorators(decoratedNode || node, source);

    // Determine kind
    let kind: SymbolKind = parentName ? 'method' : 'function';
    if (decorators.includes('property')) kind = 'property';
    if (name === '__init__') kind = 'constructor';

    // Extract parameters
    const paramsNode = getNamedChild(node, 'parameters');
    const params = paramsNode ? getNodeText(paramsNode, source) : '()';

    // Extract return type
    const returnType = this.extractReturnType(node, source);

    // Modifiers
    const modifiers: string[] = [];
    if (isAsync) modifiers.push('async');
    if (decorators.includes('staticmethod')) modifiers.push('static');
    if (decorators.includes('classmethod')) modifiers.push('classmethod');
    if (decorators.includes('abstractmethod')) modifiers.push('abstract');

    const isExported = !name.startsWith('_');
    const body = getNamedChild(node, 'block');
    const complexity = body ? this.calculateComplexity(body) : 1;
    const docComment = body ? this.extractDocstring(body, source) : null;

    symbols.push({
      name,
      kind,
      filePath,
      startLine: range.startLine,
      endLine: range.endLine,
      signature: this.buildFunctionSignature(isAsync, name, params, returnType),
      parameters: params,
      returnType,
      modifiers,
      decorators,
      parentName,
      isAsync,
      isExported,
      docComment,
      complexity,
    });

    // Extract calls from body
    if (body) {
      this.extractCalls(body, source, filePath, parentName ? `${parentName}.${name}` : name, relationships);
    }

    // Extract nested definitions
    if (body) {
      this.extractDeclarations(body, source, filePath, name, symbols, relationships);
    }

    // Decorator relationships
    for (const dec of decorators) {
      relationships.push({
        sourceSymbol: parentName ? `${parentName}.${name}` : name,
        targetSymbol: dec,
        kind: 'decorates',
        filePath,
        line: range.startLine,
      });
    }
  }

  // ─── Class Extraction ────────────────────────────────────────────────

  private extractClass(
    node: SyntaxNode, source: string, filePath: string,
    parentName: string | null, symbols: ExtractedSymbol[],
    relationships: ExtractedRelationship[],
    decoratedNode?: SyntaxNode,
  ): void {
    const nameNode = getNamedChild(node, 'identifier');
    if (!nameNode) return;

    const name = getNodeText(nameNode, source);
    const range = getNodeRange(decoratedNode || node);
    const decorators = this.getDecorators(decoratedNode || node, source);

    // Extract base classes
    const argList = getNamedChild(node, 'argument_list');
    const bases: string[] = [];
    let isProtocol = false;
    let isABC = false;

    if (argList) {
      for (let i = 0; i < argList.namedChildCount; i++) {
        const arg = argList.namedChild(i);
        if (!arg) continue;
        if (arg.type === 'identifier' || arg.type === 'attribute') {
          const baseName = getNodeText(arg, source);
          bases.push(baseName);
          if (baseName === 'Protocol') isProtocol = true;
          if (baseName === 'ABC' || baseName === 'ABCMeta') isABC = true;

          const relKind: RelationshipKind = isProtocol ? 'implements' : 'inherits';
          relationships.push({
            sourceSymbol: name,
            targetSymbol: baseName,
            kind: relKind,
            filePath,
            line: arg.startPosition.row + 1,
          });
        }
      }
    }

    const kind: SymbolKind = isProtocol ? 'interface' : 'class';
    const modifiers: string[] = [];
    if (isABC) modifiers.push('abstract');
    if (decorators.includes('dataclass')) modifiers.push('dataclass');

    const isExported = !name.startsWith('_');
    const body = getNamedChild(node, 'block');
    const docComment = body ? this.extractDocstring(body, source) : null;

    symbols.push({
      name,
      kind,
      filePath,
      startLine: range.startLine,
      endLine: range.endLine,
      signature: `class ${name}${bases.length ? `(${bases.join(', ')})` : ''}`,
      modifiers,
      decorators,
      parentName,
      isExported,
      docComment,
    });

    // Extract class body
    if (body) {
      this.extractDeclarations(body, source, filePath, name, symbols, relationships);
    }

    // Decorator relationships
    for (const dec of decorators) {
      relationships.push({
        sourceSymbol: name,
        targetSymbol: dec,
        kind: 'decorates',
        filePath,
        line: range.startLine,
      });
    }
  }

  // ─── Module Variable Extraction ──────────────────────────────────────

  private extractModuleVariable(
    node: SyntaxNode, source: string, filePath: string,
    symbols: ExtractedSymbol[],
  ): void {
    const assignment = getNamedChild(node, 'assignment');
    if (!assignment) return;

    const left = assignment.child(0);
    if (!left || left.type !== 'identifier') return;

    const name = getNodeText(left, source);
    const range = getNodeRange(node);
    const isConstant = /^[A-Z_][A-Z0-9_]*$/.test(name);

    symbols.push({
      name,
      kind: isConstant ? 'constant' as SymbolKind : 'variable',
      filePath,
      startLine: range.startLine,
      endLine: range.endLine,
      signature: getNodeText(node, source).split('\n')[0].trim().slice(0, 200),
      isExported: !name.startsWith('_'),
    });
  }

  // ─── Call Extraction ─────────────────────────────────────────────────

  private extractCalls(
    body: SyntaxNode, source: string, filePath: string,
    callerName: string, relationships: ExtractedRelationship[],
  ): void {
    const callNodes = findNodes(body, 'call');
    const seen = new Set<string>();

    for (const call of callNodes) {
      const funcNode = call.child(0);
      if (!funcNode) continue;

      const funcName = getNodeText(funcNode, source);
      // Skip common builtins
      if (['print', 'len', 'range', 'str', 'int', 'float', 'list', 'dict',
           'set', 'tuple', 'type', 'isinstance', 'issubclass', 'super',
           'hasattr', 'getattr', 'setattr', 'repr', 'bool', 'enumerate',
           'zip', 'map', 'filter', 'sorted', 'reversed', 'any', 'all',
           'min', 'max', 'abs', 'round', 'open', 'id', 'hex', 'oct',
           'bin', 'ord', 'chr', 'format', 'vars', 'dir', 'help',
           'input', 'iter', 'next', 'slice', 'object', 'property',
           'staticmethod', 'classmethod'].includes(funcName)) continue;

      const key = `${callerName}->${funcName}`;
      if (seen.has(key)) continue;
      seen.add(key);

      relationships.push({
        sourceSymbol: callerName,
        targetSymbol: funcName,
        kind: 'calls',
        filePath,
        line: call.startPosition.row + 1,
      });
    }
  }

  // ─── Helpers ─────────────────────────────────────────────────────────

  private getDecorators(node: SyntaxNode, source: string): string[] {
    const decorators: string[] = [];
    for (let i = 0; i < node.namedChildCount; i++) {
      const child = node.namedChild(i);
      if (!child || child.type !== 'decorator') continue;
      const text = getNodeText(child, source).replace(/^@/, '').split('(')[0].trim();
      decorators.push(text);
    }
    return decorators;
  }

  private extractReturnType(node: SyntaxNode, source: string): string | undefined {
    const typeNode = getNamedChild(node, 'type');
    if (typeNode) return getNodeText(typeNode, source);
    return undefined;
  }

  private extractDocstring(body: SyntaxNode, source: string): string | null {
    const firstChild = body.namedChild(0);
    if (!firstChild || firstChild.type !== 'expression_statement') return null;

    const expr = firstChild.namedChild(0);
    if (!expr || expr.type !== 'string') return null;

    const text = getNodeText(expr, source);
    return text
      .replace(/^("""|''')\s*/, '')
      .replace(/\s*("""|''')$/, '')
      .trim()
      .slice(0, 500) || null;
  }

  private calculateComplexity(node: SyntaxNode): number {
    let complexity = 1;
    const branchTypes = new Set([
      'if_statement', 'elif_clause', 'for_statement', 'while_statement',
      'except_clause', 'with_statement', 'case_clause', 'assert_statement',
    ]);

    walkTree(node, {
      enter(n) {
        if (branchTypes.has(n.type)) complexity++;
        if (n.type === 'boolean_operator') complexity++;
        if (n.type === 'conditional_expression') complexity++;
        if (['list_comprehension', 'set_comprehension',
             'dictionary_comprehension', 'generator_expression'].includes(n.type)) {
          complexity++;
        }
      }
    });

    return complexity;
  }

  private buildFunctionSignature(
    isAsync: boolean, name: string, params: string, returnType: string | undefined,
  ): string {
    const prefix = isAsync ? 'async ' : '';
    const ret = returnType ? ` -> ${returnType}` : '';
    return `${prefix}def ${name}${params}${ret}`.slice(0, 500);
  }
}
